core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java

```diff
@@ -361,8 +384,9 @@ private boolean notifyAwaitDemandFuture() {
     }
 
     private void handleCloseEvent(SubscriptionImpl subscription, CloseEvent o) {
-        setState(State.OPEN, State.CLEANUP);
-        notifySubscriberOfCloseEvent(subscription, o);
+        if (setState(State.CLOSED, State.CLEANUP)) {
```

Comment:
So the concept now is that, whatever the event which changes to the CLEANUP state, will notify the subscriber with it.